From 781e4d1d8dd1b8b6ba703f2ad540a9a33b2f4c17 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:32 -0400
Subject: [ethernet] qlcnic: Do not disable SR-IOV when VFs are assigned to VMs

Message-id: <1408626422-3941-109-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88635
O-Subject: [PATCH RHEL 7.1 108/138] qlcnic: Do not disable SR-IOV when VFs are assigned to VMs
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 696f1943a1538bb448c5bf55a18793ad410da00b

o While disabling SR-IOV when VFs are assigned to VMs causes host crash
  so return -EPERM when user request to disable SR-IOV using pci sysfs in
  case of VFs are assigned to VMs.

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
index bc85f99..3edb12f 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
@@ -454,6 +454,16 @@ static int qlcnic_pci_sriov_disable(struct qlcnic_adapter *adapter)
 {
  struct net_device *netdev = adapter->netdev;
 
+ if (pci_vfs_assigned(adapter->pdev)) {
+  netdev_err(adapter->netdev,
+      "SR-IOV VFs belonging to port %d are assigned to VMs. SR-IOV can not be disabled on this port\n",
+      adapter->portnum);
+  netdev_info(adapter->netdev,
+       "Please detach SR-IOV VFs belonging to port %d from VMs, and then try to disable SR-IOV on this port\n",
+       adapter->portnum);
+  return -EPERM;
+ }
+
  rtnl_lock();
  if (netif_running(netdev))
   __qlcnic_down(adapter, netdev);
-- 
1.7.1